src/tactic/move_add.lean

```diff
@@ -0,0 +1,224 @@
+by a `←` get placed to the left, the ones without the arrow get placed to the right.  Unnamed terms
+stay in place.  Due to re-parenthesizing, doing `move_add` with no argument may change the goal.
+
+`move_add` allows the user control over which summands should appear first and which ones
+should appear last in the rearranged expression.  Here is an example.
+
+```lean
+example (a b c d : ℕ) : a + (b + d) + c = d + a + (b + c) :=
+begin
+  move_add [a, ← b], --  Goal becomes: b + d + c + a = b + d + c + a
+  refl,
+end
+```
+
+The list of expressions that `move_add` takes is optional and a single expression can be passed
+without parentheses.  Thus `move_add ← f`, `move_add [← f]` and `move_add [←f]` all mean the same.
+
+Finally, `move_add` can also be targeted to a hypothesis.  If `hp` is in the local context,
+then `move_add [f, ← g] at hp` performs the rearranging at `hp`.
+
+##  Implementation notes
+
+The implementation of `move_add` goes via `move_add_with_rel`, which takes an extra relation
+`rel : expr → expr → bool` as a parameter.  `move_add` uses the trivial relation where all pairs
+yield true.  The presence of the relation allows, if desired, to extend the tactic to come up with
+heuristics with which to sort terms in an expression, relying less on user input.  In fact, this was
+the initial motivation for the tactic: given a polynomial, sort its terms by increasing/decreasing
+`nat_degree`.  This is no longer in scope for the current implementation of `move_add`, but could
+be developped making use of a non-trivial `rel`.
+
+Note that the tactic `abel` already implements a very solid heuristic for normalizing terms in an
+additive commutative semigroup and produces expressions in more or less standard form.
+The scope of `move_add` is different: it is designed to make it easy to move individual terms
+around a sum.
+
+##  Future work
+
+* Allow for pattern-matching in the list of terms?
+* Add support for `neg` and additive groups?
+* Add optional different operations than `+`, most notably `*`?
+* Add functionality for moving terms across the two sides of an in/dis/equality.
+* Add more tests.
+-/
+
+/--  Mimics closely `list.partition`, except that it uses the first factor as the partitioning
+condition and passes on the second factors to the partitions. -/
+def list.split_factors {α : Type*} (l : list (bool × α)) : list α × list α :=
+(l.partition (λ i : bool × α, i.1)).map (list.map (λ i, i.2)) (list.map (λ i, i.2))
+
+/--  Let `ll = (il,fl)` be a pair of `list N` and let `l` be a further list of terms from `N`.
+We move the elements of the list `l : list N`, using a relation `rel : N → N → bool`, overriding
+it as needed to make sure that the elements of the list `il` are initial and the elements of `fl`
+are final in the sorted list.
+
+Use as `l.sort_with_ends ll rel`.
+ -/
+def list.sort_with_ends {N : Type*} [decidable_eq N] (l : list N) (ll : list N × list N)
```

Comment:
I think this might be clearer as
```suggestion
def list.sort_with_ends {N : Type*} [decidable_eq N] (l : list N) (init : list N) (tail : list N)
```

Reply:
I have redesigned the pre-sorting of the lists.  This should address your comment here and also the one about the `filter` stuff.